src/azul/portal_service.py

```diff
@@ -0,0 +1,82 @@
+from copy import deepcopy
+from typing import (
+    Sequence,
+)
+import json
+import logging
+
+import boto3
+from botocore.exceptions import ClientError
+
+from azul import config
+from azul.plugin import Plugin
+from azul.types import JSON
+
+log = logging.getLogger(__name__)
+
+
+class PortalService:
+
+    def get_portal_integrations_db(self) -> Sequence[JSON]:
+        """
+        Retrieve portal DB from S3.
+
+        If no object is present at the location specified in azul.config, then
+        a default hardcoded DB is obtained from the current plugin, uploaded to
+        S3, and returned.
+        """
+
+        client = boto3.client('s3')
+        try:
+            retrieval = client.get_object(Bucket=config.portal_integrations_db_bucket,
+                                          Key=config.portal_integrations_db_object)
+        except client.exceptions.NoSuchKey:
+            plugin = Plugin.load()
+            db = self.stage_db(plugin.portal_integrations_db())
+            log.info('Portal integration DB not found in S3; using and uploading hardcoded DB from plugin.')
+            json_bytes = bytes(json.dumps(db), 'UTF-8')
+            client.put_object(Bucket=config.portal_integrations_db_bucket,
+                              Key=config.portal_integrations_db_object,
+                              Body=json_bytes,
+                              ContentType='application/json')
+        except ClientError as e:
+            log.error(f'Unexpected error retrieving portal DB from S3: {e}')
+            raise
+        else:
+            log.info("Retrieved portal integration DB from S3")
+            json_bytes = retrieval['Body'].read()
+            db = json.loads(json_bytes.decode('UTF-8'))
+        return db
+
+    def stage_db(self, unstaged_db: Sequence[JSON]) -> Sequence[JSON]:
```

Comment:
```suggestion
    def convert_hardcoded_db(self, db: Sequence[JSON]) -> Sequence[JSON]:
```

The name `stage_db` sounds like this method does the staging.

Reply:
By "unstaged" I mean "including nested entity ids for multiple deployment stages" and by "staged" I mean "including entity ids only for the current deployment stage." Are we on the same page in that regard?

Reply:
Ohh, I see. That's a rather confusing interpretation of the word _staged_ and _unstage_. 

In fact, I now regret the use of `stage` as a noun in identfiers like `dss_deployment_stage` because it leads to beauties like "the staging stage". But at the time of Azul's inception I borrowed it from the DSS team's terminology. 

I think we should stay away from "stage" alltogether. Open to suggestions. "demultiplex" and "multiplex"?

Reply:
"[de]multiplex" is satisfactory. I can't think of anything I would prefer.